File: backend/app/models/assessment.py
import json
from datetime import datetime
from uuid import UUID, uuid4

from sqlmodel import Field, SQLModel


class AssessmentResult(SQLModel, table=True):
    __tablename__ = "assessment_results"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    session_id: UUID = Field(foreign_key="sessions.id", index=True)
    child_id: UUID = Field(foreign_key="children.id", index=True)
    subject: str  # "arabic", "english", "islamic"

    # Scores by skill (JSON object)
    skill_scores: str = Field(default="{}")  # {"pronunciation": 85, "comprehension": 90}
    overall_score: int = Field(default=0)

    # Detailed feedback (JSON arrays)
    strengths: str = Field(default="[]")
    areas_for_improvement: str = Field(default="[]")
    recommendations: str = Field(default="[]")

    # Progress tracking (JSON arrays)
    mastered_skills: str = Field(default="[]")
    struggling_skills: str = Field(default="[]")

    # AI confidence and metadata
    assessment_confidence: float | None = Field(default=None)
    assessment_method: str = Field(default="rule_based")  # "rule_based", "ai_evaluated", "hybrid"

    # Time-based metrics
    response_time: int | None = Field(default=None)  # milliseconds
    time_to_complete: int | None = Field(default=None)  # minutes

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime | None = Field(default=None)

    @property
    def skill_scores_dict(self) -> dict:
        return json.loads(self.skill_scores)

    @property
    def strengths_list(self) -> list[str]:
        return json.loads(self.strengths)

    @property
    def areas_for_improvement_list(self) -> list[str]:
        return json.loads(self.areas_for_improvement)

    @property
    def recommendations_list(self) -> list[str]:
        return json.loads(self.recommendations)


class ProgressReport(SQLModel, table=True):
    __tablename__ = "progress_reports"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    child_id: UUID = Field(foreign_key="children.id", index=True)
    guardian_id: UUID = Field(foreign_key="guardians.id", index=True)

    # Report period
    start_date: datetime
    end_date: datetime

    # Overall statistics
    total_sessions: int = Field(default=0)
    total_time_spent: int = Field(default=0)  # minutes
    average_score: float = Field(default=0.0)

    # Subject-wise breakdown (JSON)
    subject_progress: str = Field(default="{}")

    # Achievements and rewards (JSON)
    badges_earned: str = Field(default="[]")
    points_earned: int = Field(default=0)

    # Parent insights (JSON arrays)
    parent_recommendations: str = Field(default="[]")
    celebrate_achievements: str = Field(default="[]")
    concern_areas: str = Field(default="[]")

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime | None = Field(default=None)

    @property
    def subject_progress_dict(self) -> dict:
        return json.loads(self.subject_progress)

    @property
    def badges_earned_list(self) -> list[dict]:
        return json.loads(self.badges_earned)

    @property
    def parent_recommendations_list(self) -> list[str]:
        return json.loads(self.parent_recommendations)

    class Config:
        validate_assignment = True

File: packages/schemas/src/assessment.ts
import { z } from "zod";
import { SubjectEnum, TimestampSchema } from "./common";

// Assessment criteria schema
export const AssessmentCriteriaSchema = z.object({
  skill: z.string(), // e.g., "pronunciation", "comprehension", "grammar"
  weight: z.number().min(0).max(1), // relative importance
  description: z.string(),
  rubric: z.array(z.object({
    level: z.enum(["novice", "developing", "proficient", "advanced"]),
    description: z.string(),
    points: z.number().min(0).max(10)
  }))
});

// Individual assessment result
export const AssessmentResultSchema = z.object({
  id: z.string().uuid(),
  sessionId: z.string().uuid(),
  childId: z.string().uuid(),
  subject: SubjectEnum,
  
  // Scores by skill
  skillScores: z.record(z.string(), z.number().min(0).max(100)),
  overallScore: z.number().min(0).max(100),
  
  // Detailed feedback
  strengths: z.array(z.string()).default([]),
  areasForImprovement: z.array(z.string()).default([]),
  recommendations: z.array(z.string()).default([]),
  
  // Progress tracking
  masteredSkills: z.array(z.string()).default([]),
  strugglingSkills: z.array(z.string()).default([]),
  
  // AI confidence and metadata
  assessmentConfidence: z.number().min(0).max(1).optional(),
  assessmentMethod: z.enum(["rule_based", "ai_evaluated", "hybrid"]),
  
  // Time-based metrics
  responseTime: z.number().optional(), // milliseconds
  timeToComplete: z.number().optional() // minutes
}).merge(TimestampSchema);

// Progress report schema
export const ProgressReportSchema = z.object({
  id: z.string().uuid(),
  childId: z.string().uuid(),
  guardianId: z.string().uuid(),
  period: z.object({
    startDate: z.string().date(),
    endDate: z.string().date()
  }),
  
  // Overall statistics
  totalSessions: z.number().min(0),
  totalTimeSpent: z.number().min(0), // minutes
  averageScore: z.number().min(0).max(100),
  
  // Subject-wise breakdown
  subjectProgress: z.record(SubjectEnum, z.object({
    sessionsCompleted: z.number().min(0),
    averageScore: z.number().min(0).max(100),
    timeSpent: z.number().min(0),
    lessonsCompleted: z.array(z.string()),
    skillsImproved: z.array(z.string()),
    nextRecommendations: z.array(z.string())
  })),
  
  // Achievements and rewards
  badgesEarned: z.array(z.object({
    badgeId: z.string(),
    earnedAt: z.string().datetime(),
    subject: SubjectEnum.optional()
  })),
  pointsEarned: z.number().min(0),
  
  // Parent insights
  parentRecommendations: z.array(z.string()).default([]),
  celebrateAchievements: z.array(z.string()).default([]),
  concernAreas: z.array(z.string()).default([])
}).merge(TimestampSchema);

// Badge/achievement schema
export const BadgeSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string(),
  iconUrl: z.string().url(),
  subject: SubjectEnum.optional(),
  criteria: z.object({
    type: z.enum(["sessions_completed", "score_threshold", "streak", "skill_mastery"]),
    value: z.number(),
    timeframe: z.enum(["day", "week", "month", "all_time"]).optional()
  }),
  rarity: z.enum(["common", "rare", "epic", "legendary"]).default("common"),
  points: z.number().min(0).default(10)
});

export type AssessmentCriteria = z.infer<typeof AssessmentCriteriaSchema>;
export type AssessmentResult = z.infer<typeof AssessmentResultSchema>;
export type ProgressReport = z.infer<typeof ProgressReportSchema>;
export type Badge = z.infer<typeof BadgeSchema>;
